### PR TITLE
[ALLUXIO-2551] Refactoring of the server module

### PR DESCRIPTION
**Pull request by jsimsa**
Addresses https://alluxio.atlassian.net/browse/ALLUXIO-2551

In particular, this PR refactors the monolithic `core-server` module into four separate modules `core-server-common`, `core-server-master`, `core-server-proxy`, and `core-server-worker`, resulting in better dependency management and the ability to reuse common server definitions without transitively inherit all dependencies of the implementation.

### CODE REVIEW

File: core/server/src/deb/control/control [hunk 1]
```diff
@@ -1,8 +0,0 @@
-Package: [[deb.pkg.name]]
```
Comment (jsimsa) [inline core/server/src/deb/control/control]: @calvinjia removing this unless you think it is needed

File: core/server/src/main/java/alluxio/worker/block/evictor/package-info.java [hunk 2]
```diff
@@ -1,49 +0,0 @@
-/*
```
Comment (jsimsa) [inline core/server/src/main/java/alluxio/worker/block/evictor/package-info.java]: @apc999 it bugs me that we only have `package-info.java` for this and one other class ... can you think of a good class to move this documentation to?

File: core/server/src/main/resources/intro.html [hunk 3]
```diff
@@ -1,12 +0,0 @@
-<!--
```
Comment (jsimsa) [inline core/server/src/main/resources/intro.html]: no longer needed

File: core/server/src/deb/control/control (hunk 1 above)
Comment (calvinjia) [inline core/server/src/deb/control/control]: Is it possible to do this in a separate PR?

File: core/server/worker/src/main/java/alluxio/web/UIFileBlockInfo.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
+/*
```
Comment (jsimsa) [inline core/server/worker/src/main/java/alluxio/web/UIFileBlockInfo.java]: this and the couple of following files need to be duplicated to avoid depending on the `core-client` from the `core-server-common`

File: core/server/worker/src/main/java/alluxio/worker/file/UnderFileSystemManager.java [hunk 5]
```diff
@@ -188,8 +188,10 @@ private InputStream openAtPosition(long position) throws IOException {
       if (mStream == null) {
         UnderFileSystem ufs = UnderFileSystem.Factory.get(mUri);
         mStream = ufs.open(mUri, OpenOptions.defaults().setOffset(position));
-      } else {
+      } else if (mStream instanceof Seekable) {
```
Comment (jsimsa) [inline core/server/worker/src/main/java/alluxio/worker/file/UnderFileSystemManager.java]: fixes checkstyle issue

File: core/server/src/deb/control/control (hunk 1 above)
Comment (jsimsa) [inline core/server/src/deb/control/control]: done

File: core/server/worker/src/main/java/alluxio/web/UIFileBlockInfo.java (hunk 4 above)
Comment (calvinjia) [inline core/server/worker/src/main/java/alluxio/web/UIFileBlockInfo.java]: Is there any way to keep the files in sync?

File: core/server/worker/src/main/java/alluxio/worker/file/UnderFileSystemManager.java [hunk 6]
```diff
@@ -188,8 +188,10 @@ private InputStream openAtPosition(long position) throws IOException {
...
         ((Seekable) mStream).seek(position);
+      } else {
+        throw new IOException("Unsupported operation.");
```
Comment (calvinjia) [inline core/server/worker/src/main/java/alluxio/worker/file/UnderFileSystemManager.java]: Could you add some detail to this message (ie. `Under File System stream does not implement seek`)? Though I think if it goes into this code block the implementation is broken. Perhaps we should enforce `mStream` is seekable? I'm fine with leaving this for another PR.

File: keyvalue/server/pom.xml [hunk 7]
```diff
@@ -46,7 +46,12 @@
     </dependency>
     <dependency>
       <groupId>org.alluxio</groupId>
-      <artifactId>alluxio-core-server</artifactId>
+      <artifactId>alluxio-core-server-master</artifactId>
+      <version>${project.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.alluxio</groupId>
+      <artifactId>alluxio-core-server-worker</artifactId>
```
Comment (calvinjia) [inline keyvalue/server/pom.xml]: We had a convention of explicitly importing everything required, should we import `common` as well?

File: core/server/common/src/main/java/alluxio/StorageTierAssoc.java [hunk 8]
```diff
@@ -21,7 +19,7 @@
 import javax.annotation.concurrent.ThreadSafe;
 
 /**
- * Creates a two-way mapping between {@link StorageTier} aliases and ordinal numbers from the given
+ * Creates a two-way mapping between storage tier aliases and ordinal numbers from the given
```
Comment (jsimsa) [inline core/server/common/src/main/java/alluxio/StorageTierAssoc.java]: yes, this class is in `core-server-common` and `StorageTier` is in `core-server-worker`

File: core/server/worker/src/main/java/alluxio/worker/file/UnderFileSystemManager.java (hunk 6 above)
Comment (jsimsa) [inline core/server/worker/src/main/java/alluxio/worker/file/UnderFileSystemManager.java]: done

File: keyvalue/server/pom.xml (hunk 7 above)
Comment (jsimsa) [inline keyvalue/server/pom.xml]: done

File: core/server/worker/src/main/java/alluxio/web/UIFileBlockInfo.java (hunk 4 above)
Comment (jsimsa) [inline core/server/worker/src/main/java/alluxio/web/UIFileBlockInfo.java]: we could put all the code that is currently duplicated code into a separate module

alternatively, we can (create a TODO to) remove the worker web UI logic that requires this code

**Comment by jsimsa** [comment]
@calvinjia I figured out how to avoid the duplication of web code ... moving `URIStatus` from `core-client` to `core-common`

File: keyvalue/server/pom.xml (hunk 7 above)
Comment (jsimsa) [inline keyvalue/server/pom.xml]: BTW, the convention is already violated in this pom as `alluxio-core-server` depends on `alluxio-core-common` and `alluxio-core-client` and neither is included in the list of dependencies

Comment (jsimsa) [inline keyvalue/server/pom.xml]: I feel like it is a difficult convention to maintain ... what are its benefits? how can we enforce it?

File: core/server/pom.xml [hunk 9]
```diff
@@ -13,14 +13,21 @@
...
   <modelVersion>4.0.0</modelVersion>
   <parent>
-    <groupId>org.alluxio</groupId>
     <artifactId>alluxio-core</artifactId>
+    <groupId>org.alluxio</groupId>
     <version>1.5.0-SNAPSHOT</version>
   </parent>
   <artifactId>alluxio-core-server</artifactId>
-  <packaging>jar</packaging>
+  <packaging>pom</packaging>
```
Comment (aaudiber) [inline core/server/pom.xml]: what is this change for?

Comment (jsimsa) [inline core/server/pom.xml]: aggregating modules should use `pom` while modules that will be used as a dependency should use `jar`

Comment (jsimsa) [inline core/server/pom.xml]: https://maven.apache.org/pom.html

**Comment by jsimsa** [comment]
@aaudiber thanks for the review ... I agree with you regarding the Hadoop dependency is, which is coming from the journaling logic ... I think our journaling should be refactored so that there is a journaling interface that can be implemented using different technologies, HDFS (and UFS in general) being one of them

File: keyvalue/server/pom.xml (hunk 7 above)
Comment (calvinjia) [inline keyvalue/server/pom.xml]: It may make less sense for internal dependencies. For external dependencies, we want to avoid the case where we remove a dependency and the lose a transitive dependency we were implicitly relying on. @apc999 do you have any thoughts on this for internal dependencies?

**Comment by calvinjia** [comment]
LGTM, are you planning to follow this up with any other changes?

**Comment by jsimsa** [comment]
@apc999 merging this now to unblock follow up work

File: keyvalue/server/pom.xml (hunk 7 above)
Comment (apc999) [inline keyvalue/server/pom.xml]: I vote for following this convention also for internal dependencies as internal dependencies may also come and go, and we still shade some modules. Here is some discussion about using direct dependencies [link](http://stackoverflow.com/questions/4662011/maven-how-do-you-deal-with-dependencies-that-are-both-direct-and-transitive/4662446#4662446).

Regarding how to enforce, `mvn dependency:tree` can help. but I would create a tool or plugin to enforce it if possible.

**Comment by apc999** [comment]
5 minutes sooner than my LGTM.

**Comment by apc999** [comment]
LGTM

File: keyvalue/server/pom.xml (hunk 7 above)
Comment (jsimsa) [inline keyvalue/server/pom.xml]: so the convention would be that all direct dependencies of a module should be in its pom? I think we need to:

1) agree on the convention for which dependencies should go into pom
2) the order in which pom dependencies should be listed in the pom
3) a tool for enforcing the convention
